src/java.base/share/classes/javax/security/auth/callback/PasswordCallback.java

```diff
@@ -107,6 +116,11 @@ public boolean isEchoOn() {
      */
     public void setPassword(char[] password) {
         this.inputPassword = (password == null ? null : password.clone());
+
+        if (this.inputPassword != null) {
+            cleanable = CleanerFactory.cleaner().register(
+                    this, cleanerFor(inputPassword));
+        }
```

Comment:
If `setPassword` is called twice in succession, should the previous password be cleaned before the new one is assigned and registered?

Reply:
I can see why that might be a good idea. Would require a specification change though. I also think it is fine to keep the behavior the same, and place the responsibility on the application to call `clearPassword` before setting a new one. We could add a warning though, something like: "Note: `clearPassword` should be called to clear any prior password before calling `setPassword` multiple times on the same `PasswordCallback` instance."

Reply:
setPassword can/should always clear the previous password.  It is an internal copy that no one else has a reference to and is being replaced.
It will need to either explicitly call cleanable.clean() or fill/erase the array itself.
Overwriting the cleanable will prevent the existing cleanable from being processed.
It don't think it needs a spec change, the internal value is implementation only.

Reply:
I think its "belt and suspenders".
If the caller does not call `clearPassword` before calling a second `setPassword,` the previous char array will still contain the previous password and remain uncleared in memory for a (longer) indeterminate time.
It is fulfilling the same purpose as the original finalizer.

Reply:
Yes, exactly. I'd recommend it calling `cleanable.clean()` prior to storing the new password, so that the cleaning action for the old password is explicitly unregistered.

Reply:
Yes, I suppose that is a good enough reason, although this class never had a finalizer AFAIK. Won't there be a small performance hit (perhaps negligible) for code that already calls `clearPassword`? A specification clarification would provide clarity to applications that they do not have to call `clearPassword` in between calls to `setPassword`. Something as simple as: "This method clears the value of any previously stored password before storing the input password".

Reply:
> Won't there be a small performance hit (perhaps negligible) for code that already calls clearPassword? 

The impact should be minimal.  If clearPassword() has been called, the cleanup (Cleanerable.clean()) won't happen again in the finalization or setPassword cleanup. 

> A specification clarification would provide clarity to applications that they do not have to call clearPassword in between calls to setPassword. 

As far as I know from the JDK code, it might be not common to call clearPassword in between calls to setPassword.  I would like to have applications calling clearPassword() methods as before, if it is not missed, to speed up the collection rather than rely on finalization.

The relationship among setPassword, getPassword and clearPassword() is complicated.  I fully agree that the spec should be clarified.  I would like to have the clarify update in another PR, and have this one focus on cleanup if an application forget to call clearPassword properly.

Reply:
> > Won't there be a small performance hit (perhaps negligible) for code that already calls clearPassword?
> 
> The impact should be minimal. If clearPassword() has been called, the cleanup (Cleanerable.clean()) won't happen again in the finalization or setPassword cleanup.

I don't think that is true, but maybe I am missing something. From looking at the code, it appears a `clearPassword` followed by a `setPassword` would call `Arrays.fill` twice on the same password. I think this can be fixed by setting the cleaner to null in the `clearPassword` method after the password has been cleared. I think this would address my concern and we may not need a spec. update (though I want to think it thru one more time).

> > A specification clarification would provide clarity to applications that they do not have to call clearPassword in between calls to setPassword.
> 
> As far as I know from the JDK code, it might be not common to call clearPassword in between calls to setPassword. I would like to have applications calling clearPassword() methods as before, if it is not missed, to speed up the collection rather than rely on finalization.

Yes, I agree calling `clearPassword` should be recommended as a best practice and callers should not solely rely on cleaners.

> The relationship among setPassword, getPassword and clearPassword() is complicated. I fully agree that the spec should be clarified. I would like to have the clarify update in another PR, and have this one focus on cleanup if an application forget to call clearPassword properly.

See above.

Reply:
Calling `Cleanable.clean()` calls the `Runnable` action at-most-once.
Each `Cleanable` inserted in a list when it is created and is removed when `clear()` or `clean()` is invoked.
The action is called only if it is still in the list. So there is no extra overhead.

There is no harm in clearing the cleanable field; but it does not save much.

The code in `clearPassword` can be simplified and only test `cleanable != null`; it will be null unless there is an inputPassword to clean.

Reply:
I'd recommend setting `cleanable` to null after it's been cleaned to make the state machine easier to reason about. The invariant would be: if `cleanable` is non-null, then we have something dirty that needs to be cleaned. If we don't clear it to null after cleaning, it potentially results in confusing states. For example, suppose the app calls `setPassword(nonNull)` and later calls `setPassword(null)`. The second call will set `inputPassword` to null but leave a stale reference in `cleanable`. This isn't necessarily harmful, but it's confusing.

Reply:
> The code in `clearPassword` can be simplified and only test `cleanable != null`; it will be null unless there is an inputPassword to clean.

Yes.  The testing of `cleanable != null` is sufficient.

Reply:
> I'd recommend setting `cleanable` to null after it's been cleaned to make the state machine easier to reason about.

I like this idea.  Updated.